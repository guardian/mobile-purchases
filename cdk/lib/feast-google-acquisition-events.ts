import type { GuStackProps } from '@guardian/cdk/lib/constructs/core';
import { GuStack } from "@guardian/cdk/lib/constructs/core";
import { GuLambdaFunction } from '@guardian/cdk/lib/constructs/lambda/lambda';
import type { App } from 'aws-cdk-lib';
import { Duration } from 'aws-cdk-lib';
import {
	Effect,
	Policy,
	PolicyStatement,
} from 'aws-cdk-lib/aws-iam';
import { Runtime } from 'aws-cdk-lib/aws-lambda';
import { SqsEventSource } from 'aws-cdk-lib/aws-lambda-event-sources';
import { Queue } from 'aws-cdk-lib/aws-sqs';

export interface FeastGoogleAcquisitionEventsProps extends GuStackProps {
	stack: string;
	stage: string;
}

export class FeastGoogleAcquisitionEvents extends GuStack {
	constructor(scope: App, id: string, props: FeastGoogleAcquisitionEventsProps) {
		super(scope, id, props);

		const app = 'feast-google-acquisition-events';
		const nameWithStage = `${app}-${this.stage}`;

		const commonEnvironmentVariables = {
			App: app,
			Stack: this.stack,
			Stage: this.stage,
		};

		const nodeVersion = Runtime.NODEJS_20_X;

		//SQS
		const queueName = `${app}-queue-${props.stage}`;
		const deadLetterQueueName = `${app}-dlq-${props.stage}`;

		const deadLetterQueue = new Queue(this, deadLetterQueueName, {
			queueName: deadLetterQueueName,
			retentionPeriod: Duration.days(14),
		});

		const queue = new Queue(this, queueName, {
			queueName,
			deadLetterQueue: {
				// The number of times a message can be unsuccessfully dequeued before being moved to the dlq
				maxReceiveCount: 5,
				queue: deadLetterQueue,
			},
			// This must be >= the lambda timeout
			visibilityTimeout: Duration.minutes(5),
		});

		// SQS to Lambda event source mapping
		const eventSource = new SqsEventSource(queue, {
			reportBatchItemFailures: true,
		});
		const events = [eventSource];

		// ---- API-triggered lambda functions ---- //
		const lambda = new GuLambdaFunction(this, `${app}-lambda`, {
			description:
				'An API Gateway triggered lambda generated in the support-service-lambdas repo',
			functionName: nameWithStage,
			fileName: `${app}.zip`,
			handler: 'index.handler',
			runtime: nodeVersion,
			memorySize: 1024,
			timeout: Duration.seconds(300),
			environment: commonEnvironmentVariables,
			app: app,
			events,
		});

		const s3InlinePolicy: Policy = new Policy(this, 'S3 inline policy', {
			statements: [
				new PolicyStatement({
					effect: Effect.ALLOW,
					actions: ['s3:GetObject'],
					resources: [
						`arn:aws:s3::*:mobile-dist/${this.stack}/${this.stage}/${app}/`,
					],
				}),
			],
		});

		lambda.role?.attachInlinePolicy(s3InlinePolicy);
	}
}