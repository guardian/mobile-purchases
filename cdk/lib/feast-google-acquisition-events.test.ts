import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import {FeastGoogleAcquisitionEvents} from "./feast-google-acquisition-events";

describe('The FeastGoogleAcquisitionEvents stack', () => {
    it('matches the snapshot', () => {
        const app = new App();
        const codeStack = new FeastGoogleAcquisitionEvents(app, 'feast-google-acquisition-events-CODE',{ stack: 'mobile', stage: 'CODE' });
        const prodStack = new FeastGoogleAcquisitionEvents(app, 'feast-google-acquisition-events-PROD',{ stack: 'mobile', stage: 'PROD' });

        expect(Template.fromStack(codeStack).toJSON()).toMatchSnapshot();
        expect(Template.fromStack(prodStack).toJSON()).toMatchSnapshot();
    });
});