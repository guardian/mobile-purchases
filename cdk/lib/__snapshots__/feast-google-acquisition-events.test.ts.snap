// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The FeastGoogleAcquisitionEvents stack matches the snapshot 1`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [
      "GuDistributionBucketParameter",
      "GuLambdaFunction",
    ],
    "gu:cdk:version": "TEST",
  },
  "Parameters": {
    "DistributionBucketName": {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "S3inlinepolicy3B07399A": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Resource": "arn:aws:s3::*:mobile-dist/mobile/CODE/feast-google-acquisition-events/",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "S3inlinepolicy3B07399A",
        "Roles": [
          {
            "Ref": "feastgoogleacquisitioneventslambdaServiceRoleAB27E589",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "feastgoogleacquisitioneventsdlqCODEAE494D5F": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "MessageRetentionPeriod": 1209600,
        "QueueName": "feast-google-acquisition-events-dlq-CODE",
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/mobile-purchases",
          },
          {
            "Key": "Stack",
            "Value": "mobile",
          },
          {
            "Key": "Stage",
            "Value": "CODE",
          },
        ],
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "feastgoogleacquisitioneventslambda2B309244": {
      "DependsOn": [
        "feastgoogleacquisitioneventslambdaServiceRoleDefaultPolicy1A3A64B5",
        "feastgoogleacquisitioneventslambdaServiceRoleAB27E589",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "DistributionBucketName",
          },
          "S3Key": "mobile/CODE/feast-google-acquisition-events/feast-google-acquisition-events.zip",
        },
        "Description": "An API Gateway triggered lambda generated in the support-service-lambdas repo",
        "Environment": {
          "Variables": {
            "APP": "feast-google-acquisition-events",
            "App": "feast-google-acquisition-events",
            "STACK": "mobile",
            "STAGE": "CODE",
            "Stack": "mobile",
            "Stage": "CODE",
          },
        },
        "FunctionName": "feast-google-acquisition-events-CODE",
        "Handler": "index.handler",
        "MemorySize": 1024,
        "Role": {
          "Fn::GetAtt": [
            "feastgoogleacquisitioneventslambdaServiceRoleAB27E589",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Tags": [
          {
            "Key": "App",
            "Value": "feast-google-acquisition-events",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/mobile-purchases",
          },
          {
            "Key": "Stack",
            "Value": "mobile",
          },
          {
            "Key": "Stage",
            "Value": "CODE",
          },
        ],
        "Timeout": 300,
      },
      "Type": "AWS::Lambda::Function",
    },
    "feastgoogleacquisitioneventslambdaServiceRoleAB27E589": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": [
          {
            "Key": "App",
            "Value": "feast-google-acquisition-events",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/mobile-purchases",
          },
          {
            "Key": "Stack",
            "Value": "mobile",
          },
          {
            "Key": "Stage",
            "Value": "CODE",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "feastgoogleacquisitioneventslambdaServiceRoleDefaultPolicy1A3A64B5": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "sqs:ReceiveMessage",
                "sqs:ChangeMessageVisibility",
                "sqs:GetQueueUrl",
                "sqs:DeleteMessage",
                "sqs:GetQueueAttributes",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "feastgoogleacquisitioneventsqueueCODE1125D8B7",
                  "Arn",
                ],
              },
            },
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                      "/mobile/CODE/feast-google-acquisition-events/feast-google-acquisition-events.zip",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/CODE/mobile/feast-google-acquisition-events",
                  ],
                ],
              },
            },
            {
              "Action": [
                "ssm:GetParameters",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/CODE/mobile/feast-google-acquisition-events/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "feastgoogleacquisitioneventslambdaServiceRoleDefaultPolicy1A3A64B5",
        "Roles": [
          {
            "Ref": "feastgoogleacquisitioneventslambdaServiceRoleAB27E589",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "feastgoogleacquisitioneventslambdaSqsEventSourcefeastgoogleacquisitioneventsCODEfeastgoogleacquisitioneventsqueueCODEF024DC5AE08E681C": {
      "Properties": {
        "EventSourceArn": {
          "Fn::GetAtt": [
            "feastgoogleacquisitioneventsqueueCODE1125D8B7",
            "Arn",
          ],
        },
        "FunctionName": {
          "Ref": "feastgoogleacquisitioneventslambda2B309244",
        },
        "FunctionResponseTypes": [
          "ReportBatchItemFailures",
        ],
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
    "feastgoogleacquisitioneventsqueueCODE1125D8B7": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "QueueName": "feast-google-acquisition-events-queue-CODE",
        "RedrivePolicy": {
          "deadLetterTargetArn": {
            "Fn::GetAtt": [
              "feastgoogleacquisitioneventsdlqCODEAE494D5F",
              "Arn",
            ],
          },
          "maxReceiveCount": 5,
        },
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/mobile-purchases",
          },
          {
            "Key": "Stack",
            "Value": "mobile",
          },
          {
            "Key": "Stage",
            "Value": "CODE",
          },
        ],
        "VisibilityTimeout": 300,
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;

exports[`The FeastGoogleAcquisitionEvents stack matches the snapshot 2`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [
      "GuDistributionBucketParameter",
      "GuLambdaFunction",
    ],
    "gu:cdk:version": "TEST",
  },
  "Parameters": {
    "DistributionBucketName": {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "S3inlinepolicy3B07399A": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Resource": "arn:aws:s3::*:mobile-dist/mobile/PROD/feast-google-acquisition-events/",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "S3inlinepolicy3B07399A",
        "Roles": [
          {
            "Ref": "feastgoogleacquisitioneventslambdaServiceRoleAB27E589",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "feastgoogleacquisitioneventsdlqPROD0BD7F8C4": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "MessageRetentionPeriod": 1209600,
        "QueueName": "feast-google-acquisition-events-dlq-PROD",
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/mobile-purchases",
          },
          {
            "Key": "Stack",
            "Value": "mobile",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "feastgoogleacquisitioneventslambda2B309244": {
      "DependsOn": [
        "feastgoogleacquisitioneventslambdaServiceRoleDefaultPolicy1A3A64B5",
        "feastgoogleacquisitioneventslambdaServiceRoleAB27E589",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "DistributionBucketName",
          },
          "S3Key": "mobile/PROD/feast-google-acquisition-events/feast-google-acquisition-events.zip",
        },
        "Description": "An API Gateway triggered lambda generated in the support-service-lambdas repo",
        "Environment": {
          "Variables": {
            "APP": "feast-google-acquisition-events",
            "App": "feast-google-acquisition-events",
            "STACK": "mobile",
            "STAGE": "PROD",
            "Stack": "mobile",
            "Stage": "PROD",
          },
        },
        "FunctionName": "feast-google-acquisition-events-PROD",
        "Handler": "index.handler",
        "MemorySize": 1024,
        "Role": {
          "Fn::GetAtt": [
            "feastgoogleacquisitioneventslambdaServiceRoleAB27E589",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Tags": [
          {
            "Key": "App",
            "Value": "feast-google-acquisition-events",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/mobile-purchases",
          },
          {
            "Key": "Stack",
            "Value": "mobile",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
        "Timeout": 300,
      },
      "Type": "AWS::Lambda::Function",
    },
    "feastgoogleacquisitioneventslambdaServiceRoleAB27E589": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": [
          {
            "Key": "App",
            "Value": "feast-google-acquisition-events",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/mobile-purchases",
          },
          {
            "Key": "Stack",
            "Value": "mobile",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "feastgoogleacquisitioneventslambdaServiceRoleDefaultPolicy1A3A64B5": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "sqs:ReceiveMessage",
                "sqs:ChangeMessageVisibility",
                "sqs:GetQueueUrl",
                "sqs:DeleteMessage",
                "sqs:GetQueueAttributes",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "feastgoogleacquisitioneventsqueuePROD8E7ABE58",
                  "Arn",
                ],
              },
            },
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                      "/mobile/PROD/feast-google-acquisition-events/feast-google-acquisition-events.zip",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/PROD/mobile/feast-google-acquisition-events",
                  ],
                ],
              },
            },
            {
              "Action": [
                "ssm:GetParameters",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/PROD/mobile/feast-google-acquisition-events/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "feastgoogleacquisitioneventslambdaServiceRoleDefaultPolicy1A3A64B5",
        "Roles": [
          {
            "Ref": "feastgoogleacquisitioneventslambdaServiceRoleAB27E589",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "feastgoogleacquisitioneventslambdaSqsEventSourcefeastgoogleacquisitioneventsPRODfeastgoogleacquisitioneventsqueuePRODCEE4A3E5200426E4": {
      "Properties": {
        "EventSourceArn": {
          "Fn::GetAtt": [
            "feastgoogleacquisitioneventsqueuePROD8E7ABE58",
            "Arn",
          ],
        },
        "FunctionName": {
          "Ref": "feastgoogleacquisitioneventslambda2B309244",
        },
        "FunctionResponseTypes": [
          "ReportBatchItemFailures",
        ],
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
    "feastgoogleacquisitioneventsqueuePROD8E7ABE58": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "QueueName": "feast-google-acquisition-events-queue-PROD",
        "RedrivePolicy": {
          "deadLetterTargetArn": {
            "Fn::GetAtt": [
              "feastgoogleacquisitioneventsdlqPROD0BD7F8C4",
              "Arn",
            ],
          },
          "maxReceiveCount": 5,
        },
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/mobile-purchases",
          },
          {
            "Key": "Stack",
            "Value": "mobile",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
        "VisibilityTimeout": 300,
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;
